### PR TITLE
Error reporting

### PR DESCRIPTION
**Pull request by Ichbinjoe**
Addresses #15 ~~currently without internal server exceptions (only plugin caused exceptions)~~

TODO: 
- ~~Plugin Messaging Handling (I forgot...)~~
- ~~Internal~~

### CODE REVIEW

**Comment by Ichbinjoe** [comment]
I got the entity and tile entity based crash saves.. where are the other internal exception reporting points?

**Comment by zachbr** [comment]
I'm not sure what your thought process is on what exceptions in NMS and OBC are worth calling for, but we are missing a substantial number of internal exceptions.
Running `Find in Path...` on PaperSpigot-Server reveals many of these, not all worth calling for obviously, but many many more places where this should be considered

**Comment by Ichbinjoe** [comment]
Yeah, I didn't look overly hard. I thought there were some methods I missed (forgot about them :P) Basically my whole thought is that if it can be recovered from, it should be reported. I'll look more heavily in a bit.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+         }
+     }
+diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+index 1325b03..ebf2226 100644
+--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
++++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+@@ -1,23 +1,6 @@
+ package org.bukkit.plugin;
+ 
+-import java.io.File;
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Don't go changing imports like this

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+                 server.getLogger().log(Level.SEVERE, "Error occurred (in the plugin loader) while enabling " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
++                handlePluginException(ex, plugin);
+             }
+ 
+             HandlerList.bakeAll();
+@@ -423,24 +416,28 @@ public final class SimplePluginManager implements PluginManager {
+                 plugin.getPluginLoader().disablePlugin(plugin);
+             } catch (Throwable ex) {
+                 server.getLogger().log(Level.SEVERE, "Error occurred (in the plugin loader) while disabling " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
++                handlePluginException(ex,plugin); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Formatting

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+             }
+ 
+             try {
+                 server.getScheduler().cancelTasks(plugin);
+             } catch (Throwable ex) {
+                 server.getLogger().log(Level.SEVERE, "Error occurred (in the plugin loader) while cancelling tasks for " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
++                handlePluginException(ex,plugin); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Formatting

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+             }
+ 
+             try {
+                 server.getServicesManager().unregisterAll(plugin);
+             } catch (Throwable ex) {
+                 server.getLogger().log(Level.SEVERE, "Error occurred (in the plugin loader) while unregistering services for " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
++                handlePluginException(ex,plugin); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Formatting

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+             }
+ 
+             try {
+                 HandlerList.unregisterAll(plugin);
+             } catch (Throwable ex) {
+                 server.getLogger().log(Level.SEVERE, "Error occurred (in the plugin loader) while unregistering events for " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
++                handlePluginException(ex,plugin); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Formatting

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+             }
+ 
+             try {
+@@ -448,10 +445,17 @@ public final class SimplePluginManager implements PluginManager {
+                 server.getMessenger().unregisterOutgoingPluginChannel(plugin);
+             } catch(Throwable ex) {
+                 server.getLogger().log(Level.SEVERE, "Error occurred (in the plugin loader) while unregistering plugin channels for " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
++                handlePluginException(ex,plugin); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Formatting

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 7]
```diff
@@ -0,0 +1,605 @@
...
++import org.bukkit.command.Command;
++import org.bukkit.command.CommandSender;
++
++/**
++ * Thrown when a command throws an exception
++ */
++public class ServerCommandException extends ServerException {
++    public ServerCommandException(String message, Throwable cause, Command command, CommandSender commandSender, String[] arguments) {
++        super(message, cause);
++        Validate.notNull(command);
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: If you use Guava's `Preconditions.checkNotNull` method you can do:
`this.command = checkNotNull(command, "command");`

(applies to all cases of this)

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
++        super(cause, task.getOwner());
++        this.task = task;
++    }
++
++    protected ServerSchedulerException(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace, BukkitTask task) {
++        super(message, cause, enableSuppression, writableStackTrace, task.getOwner());
++        this.task = task;
++    }
++
++    private final BukkitTask task;
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: This should be at the top of the class to be consistent.

File: Spigot-Server-Patches/0097-Add-exception-reporting-event.patch [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
++import org.github.paperspigot.event.ServerExceptionEvent;
++import org.github.paperspigot.exception.ServerInternalException;
+ // PaperSpigot end
+ 
+ // CraftBukkit start
+@@ -1452,6 +1454,7 @@ public abstract class World implements IBlockAccess {
+                     entity.tickTimer.stopTiming();
+                     System.err.println("Entity threw exception at " + entity.world.getWorld().getName() + ":" + entity.locX + "," + entity.locY + "," + entity.locZ);
+                     throwable1.printStackTrace();
++                    getServer().getPluginManager().callEvent(new ServerExceptionEvent(new ServerInternalException(throwable1)));
```
Comment (kashike) [inline Spigot-Server-Patches/0097-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-Server-Patches/0097-Add-exception-reporting-event.patch [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                     entity.dead = true;
+                     continue;
+                     // PaperSpigot end
+@@ -1515,6 +1518,7 @@ public abstract class World implements IBlockAccess {
+                         tileentity.tickTimer.stopTiming();
+                         System.err.println("TileEntity threw exception at " + tileentity.world.getWorld().getName() + ":" + tileentity.position.getX() + "," + tileentity.position.getY() + "," + tileentity.position.getZ());
+                         throwable2.printStackTrace();
++                        getServer().getPluginManager().callEvent(new ServerExceptionEvent(new ServerInternalException(throwable2)));
```
Comment (kashike) [inline Spigot-Server-Patches/0097-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-Server-Patches/0097-Add-exception-reporting-event.patch [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+ 
+ /**
+  * The fundamental concepts for this implementation:
+@@ -355,11 +358,16 @@ public class CraftScheduler implements BukkitScheduler {
+                                 task.getTaskId(),
+                                 task.getOwner().getDescription().getFullName()),
+                             throwable);
++                    // Paper start
++                    task.getOwner().getServer().getPluginManager().callEvent(
++                            new ServerExceptionEvent(new ServerSchedulerException(throwable, task))
```
Comment (kashike) [inline Spigot-Server-Patches/0097-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
++import org.github.paperspigot.exception.ServerCommandException;
++import org.github.paperspigot.exception.ServerTabCompleteException;
+ 
+ public class SimpleCommandMap implements CommandMap {
+     private static final Pattern PATTERN_ON_SPACE = Pattern.compile(" ", Pattern.LITERAL);
+@@ -147,6 +150,7 @@ public class SimpleCommandMap implements CommandMap {
+             throw ex;
+         } catch (Throwable ex) {
+             target.timings.stopTiming(); // Spigot
++            server.getPluginManager().callEvent(new ServerExceptionEvent(new ServerCommandException(ex, target, sender, args))); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch:27]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 13]
```diff
@@ -0,0 +1,605 @@
...
+             throw new CommandException("Unhandled exception executing '" + commandLine + "' in " + target, ex);
+         }
+ 
+@@ -224,6 +228,7 @@ public class SimpleCommandMap implements CommandMap {
+         } catch (CommandException ex) {
+             throw ex;
+         } catch (Throwable ex) {
++            server.getPluginManager().callEvent(new ServerExceptionEvent(new ServerTabCompleteException(ex, target, sender, args))); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
++import java.util.regex.Matcher;
++import java.util.regex.Pattern;
+ 
+ /**
+  * Handles all plugin management from the Server
+@@ -404,6 +396,7 @@ public final class SimplePluginManager implements PluginManager {
+                 plugin.getPluginLoader().enablePlugin(plugin);
+             } catch (Throwable ex) {
+                 server.getLogger().log(Level.SEVERE, "Error occurred (in the plugin loader) while enabling " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
++                handlePluginException(ex, plugin);
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch (hunk 2 above)
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch (hunk 3 above)
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch (hunk 4 above)
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch (hunk 5 above)
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch (hunk 6 above)
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+     public void clearPlugins() {
+         synchronized (this) {
+             disablePlugins();
+@@ -514,6 +518,10 @@ public final class SimplePluginManager implements PluginManager {
+                 }
+             } catch (Throwable ex) {
+                 server.getLogger().log(Level.SEVERE, "Could not pass event " + event.getEventName() + " to " + registration.getPlugin().getDescription().getFullName(), ex);
++                // Paper start - error reporting
++                if (!(event instanceof ServerExceptionEvent)) { // We don't want to cause an endless event loop
++                    callEvent(new ServerExceptionEvent(new ServerEventException(ex, registration.getPlugin(), registration.getListener(), event)));
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: Message associated with the exception might be handy.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 16]
```diff
@@ -0,0 +1,605 @@
...
++    }
++
++    protected ServerPluginMessageException(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace, Plugin responsiblePlugin, Player player, String channel, byte[] data) {
++        super(message, cause, enableSuppression, writableStackTrace, responsiblePlugin);
++        this.player = player;
++        this.channel = channel;
++        this.data = data;
++    }
++
++    private final Player player;
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: This should be at the top of the class to be consistent.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 17]
```diff
@@ -0,0 +1,605 @@
...
++        this.responsiblePlugin = responsiblePlugin;
++    }
++
++    protected ServerPluginException(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace, Plugin responsiblePlugin) {
++        super(message, cause, enableSuppression, writableStackTrace);
++        Validate.notNull(responsiblePlugin);
++        this.responsiblePlugin = responsiblePlugin;
++    }
++
++    private final Plugin responsiblePlugin;
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch:463]: This should be at the top of the class to be consistent.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 18]
```diff
@@ -0,0 +1,605 @@
...
++
++    protected ServerEventException(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace, Plugin responsiblePlugin, Listener listener, Event event) {
++        super(message, cause, enableSuppression, writableStackTrace, responsiblePlugin);
++        Validate.notNull(listener);
++        Validate.notNull(event);
++        this.listener = listener;
++        this.event = event;
++    }
++
++    private final Listener listener;
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: This should be at the top of the class to be consistent.

File: Spigot-API-Patches/0021-Add-exception-reporting-event.patch [hunk 19]
```diff
@@ -0,0 +1,605 @@
...
++        super(message, cause, enableSuppression, writableStackTrace);
++        Validate.notNull(command);
++        Validate.notNull(commandSender);
++        Validate.notNull(arguments);
++        this.command = command;
++        this.commandSender = commandSender;
++        this.arguments = arguments;
++    }
++
++    private final Command command;
```
Comment (kashike) [inline Spigot-API-Patches/0021-Add-exception-reporting-event.patch]: This should be at the top of the class to be consistent.

File: Spigot-Server-Patches/0097-Add-exception-reporting-event.patch [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
++            );
++            throw e;
++        } catch (Throwable t) {
++            internalTask.getOwner().getServer().getPluginManager().callEvent(
++                    new ServerExceptionEvent(new ServerSchedulerException(t, internalTask))
++            ); //Do not rethrow, since it is not permitted with Runnable#run
++        }
++    }
++
++    private final CraftTask internalTask;
```
Comment (kashike) [inline Spigot-Server-Patches/0097-Add-exception-reporting-event.patch]: This should be at the top of the class to be consistent.

**Comment by kashike** [comment]
What's with the commits?

**Comment by Ichbinjoe** [comment]
I wanted all of my code base I was editing on top of to be up to date, as well as the patch numbers to be in order. I wasn't aware Github was going to freak out like this... Just thought it would update the branch :P

**Pull request merged**